stm32: get bytes left to transfer via DMA and set transfer complete callback

seeseemelk opened the pull request:
### Contribution description
This PR adds two additional functions to the STM32 DMA API:
 - `dma_left(dma)` - Gets the number of bytes that still have to be transferred by the DMA controller. This allows retrieving this number without having to suspend and resume the transfer using `dma_suspend` and `dma_resume`.
 - `dma_transfer_callback(dma, callback, arg)` - Allows setting a user-defined callback which gets executed when the Transfer Complete interrupt is fired.

benpicco commented on cpu/stm32/include/periph_cpu.h:
```diff
@@ -965,6 +965,15 @@ void dma_start(dma_t dma);
+ *
+ * @return the remaining number of bytes to transfer.
+ */
+uint16_t dma_left(dma_t dma);
```
This is inherently racy - what's the use case for this?

benpicco commented on cpu/stm32/periph/dma.c:
```diff
@@ -445,6 +447,13 @@ int dma_configure(dma_t dma, int chan, const volatile void *src, volatile void *
     return 0;
 }
 
+void dma_transfer_complete(dma_t dma, void (*callback)(void*, dma_t), void *arg)
```
I'd go with something less ambigious

```suggestion
void dma_set_transfer_complete_cb(dma_t dma, void (*callback)(void*, dma_t), void *arg)
```

benpicco commented on cpu/stm32/periph/dma.c:
```diff
@@ -445,6 +447,13 @@ int dma_configure(dma_t dma, int chan, const volatile void *src, volatile void *
+void dma_transfer_complete(dma_t dma, void (*callback)(void*, dma_t), void *arg)
+{
+    struct dma_ctx *ctx = &dma_ctx[dma];
+    ctx->callback = callback;
+    ctx->callback_arg = arg;
```
```suggestion
    unsigned state = irq_disable();
    ctx->callback = callback;
    ctx->callback_arg = arg;
    irq_restore(state);
```

benpicco commented on cpu/stm32/periph/dma.c:
```diff
@@ -504,7 +521,11 @@ void dma_isr_handler(dma_t dma)
 
-    mutex_unlock(&dma_ctx[dma].sync_lock);
+    struct dma_ctx *ctx = &dma_ctx[dma];
+    mutex_unlock(&ctx->sync_lock);
```
I guess we could replace this by setting `ctx->callback = mutex_unlock` now.

benpicco reviewed (commented):
The usual way would be to `dma_wait()` and then execute whatever is supposed to happen after the DMA completion following that.

Is this not possible due to timing constraints?

seeseemelk commented:
The scenario I'm in is the following: I'm expecting to approximately half a megabyte of data over UART at a baudrate of 1000000 (that's 1 MHz, not 100 KHz). As that is more data than I have memory available, I have to process it as it's coming in.

With RIOT's current API I'm having two issues:
 - The data I receive is of unknown size. It might be a nice multiply of my buffer, or it might not. Waiting for all data to be received using `dma_wait` could block forever, getting the bytes left using `dma_suspend` would pause the transfer, potentially resulting in data loss.
 - The external peripheral should not have to wait for the DMA controller to be reconfigured when it is full.

In order to work around the first issue, I added `dma_left`. This allows me to known how much data is already in the buffer (by taking the size of the buffer and subtracting the result of `dma_left`). Based on the contents of the received data, I can know whether I've reached the end or not.

In order to fix the second issue I added a function which lets me execute a callback when my receive buffer is full. I can then reconfigure the DMA controller to start writing back into the start of the receive buffer up to the read pointer. This essentially implements a circular buffer, but with the difference that it will stop the flow of data when my read pointer has not yet caught up (this happens when I do a lot of logging over a 115200 baud UART)

I hope this makes it a little clearer why I would like these functions, and I will definitily be fixing some of the race conditions I created in here tomorrow.
